src/Transformer/DetermineHeroImages.php

```diff
@@ -68,44 +68,46 @@ final class DetermineHeroImages implements Transformer {
 	 * @return void
 	 */
 	public function transform( Document $document, ErrorCollection $errors ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
-		$existing_hero_images_count = $document->xpath->query(
-			self::EXISTING_HERO_IMAGES_XPATH_QUERY,
-			$document->body
-		)->length;
-
-		$available_hero_image_slots = max(
-			PreloadHeroImage::DATA_HERO_MAX - $existing_hero_images_count,
-			0
-		);
-
 		$hero_image_elements = [];
 
-		if ( count( $hero_image_elements ) < $available_hero_image_slots ) {
-			$custom_logo = $this->get_custom_logo( $document );
-			if ( null !== $custom_logo ) {
-				$hero_image_elements[] = $custom_logo;
-			}
-		}
+		foreach ( [ 'custom_header', 'custom_logo', 'featured_image', 'cover_blocks' ] as $hero_image_source ) {
+			if ( count( $hero_image_elements ) < PreloadHeroImage::DATA_HERO_MAX ) {
+				$get_candidate_method = "get_{$hero_image_source}";
+				$candidates           = $this->$get_candidate_method( $document );
+
+				if ( null === $candidates ) {
+					continue;
+				}
 
-		if ( count( $hero_image_elements ) < $available_hero_image_slots ) {
-			$featured_image = $this->get_featured_image( $document );
-			if ( null !== $featured_image ) {
-				$hero_image_elements[] = $featured_image;
+				if ( is_array( $candidates ) ) {
+					$hero_image_elements = array_merge( $hero_image_elements, array_filter( $candidates ) );
+				} else {
+					$hero_image_elements[] = $candidates;
+				}
 			}
 		}
 
-		if ( count( $hero_image_elements ) < $available_hero_image_slots ) {
-			$hero_image_elements = array_merge(
-				$hero_image_elements,
-				array_filter(
-					$this->get_cover_blocks( $document )
-				)
-			);
-		}
+		$this->add_data_hero_candidate_attribute(
+			array_slice( $hero_image_elements, 0, PreloadHeroImage::DATA_HERO_MAX )
+		);
+	}
 
-		$this->add_data_hero_attribute(
-			array_slice( $hero_image_elements, 0, $available_hero_image_slots )
+	/**
+	 * Retrieve the element that represents the custom header.
+	 *
+	 * @param Document $document Document to retrieve the custom header from.
+	 * @return DOMElement|null Element that represents the custom header, or null
+	 *                         if not found.
+	 */
+	private function get_custom_header( Document $document ) {
+		$elements = $document->xpath->query(
+			self::CUSTOM_HEADER_XPATH_QUERY,
+			$document->body
 		);
```

Comment:
An alternative way to identify the header image which may be more robust as opposed to relying on element IDs is to obtain the header image URL and then query for image elements that have that `src`:

```suggestion
		$header_image = get_header_image();
		if ( ! $header_image ) {
			return null;
		}

		$elements = $document->xpath->query(
			sprintf(
				'//*[ ( self::img or self::amp-img ) and @src = "%s" ]',
				addcslashes( $custom_header->url, '"\\' )
			),
			$document->body
		);
```

But this is going to have other problems, namely because `get_header_image_tag()` allows for the markup to be filtered, meaning the image URL could be replaced with one that is pointing to a CDN.